University of California Press Books.js

```diff
@@ -0,0 +1,403 @@
+{
+	"translatorID": "f27b5031-01fb-4f14-8829-6748f819eac4",
+	"label": "University of California Press Books",
+	"creator": "Abe Jellinek",
+	"target": "^https://www\\.ucpress\\.edu/(book/|search.php\\?)",
+	"minVersion": "3.0",
+	"maxVersion": "",
+	"priority": 100,
+	"inRepository": true,
+	"translatorType": 4,
+	"browserSupport": "gcsibv",
+	"lastUpdated": "2021-06-29 06:45:51"
+}
+
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2021 Abe Jellinek
+	
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+
+function detectWeb(doc, url) {
+	if (url.includes('/book/')) {
+		return "book";
+	}
+	else if (getSearchResults(doc, true)) {
+		return "multiple";
+	}
+	return false;
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	var rows = doc.querySelectorAll('a.gs-title[href*="/book/"]');
+	for (let row of rows) {
+		let href = row.href;
+		let title = ZU.trimInternal(row.textContent);
+		if (!href || !title) continue;
+		if (checkOnly) return true;
+		found = true;
+		items[href] = title;
+	}
+	return found ? items : false;
+}
+
+function doWeb(doc, url) {
+	if (detectWeb(doc, url) == "multiple") {
+		Zotero.selectItems(getSearchResults(doc, false), function (items) {
+			if (items) ZU.processDocuments(Object.keys(items), scrape);
+		});
+	}
+	else {
+		scrape(doc, url);
+	}
+}
+
+function scrape(doc, url) {
+	var translator = Zotero.loadTranslator('web');
+	// Embedded Metadata
+	translator.setTranslator('951c027d-74ac-47d4-a107-9c3069ab7b48');
```

Comment:
Hi Abe, sorry for non-review-related comment--I was just browsing to learn how to write better translators. Please feel free to ignore my question: I see this line here and there but never understood what function it does. Could you explain shortly or point to documentation? Thx!

Reply:
For sure! `Zotero.loadTranslator` returns a `Translate` object, which is essentially a driver for the translation system. Passing `'web'` sets it up to run web translators, but it doesn't know *which* translator to run yet. That's what `Translate.setTranslator` is for. You pass in a translator's unique ID, as in the metadata block:

https://github.com/zotero/translators/blob/7996f2ede84b514023d2f55d0ec6343e3160e4a3/Embedded%20Metadata.js#L2

In this case we want EM, so we pass its ID. Calling `Translate.getTranslatorObject` gives us an instance of EM with `detectWeb` and `doWeb` methods that we can call.

Loading search translators looks a bit different:

https://github.com/zotero/translators/blob/d147639248c81830f727928d9b7256edee05d306/COinS.js#L95-L120

Reply:
Take a look at the [type definition file](https://github.com/zotero/translators/blob/master/index.d.ts) in this repo if you're ever wondering what returns what!

Reply:
I see! I didn't know there are these great generic translators we could call. Thanks so much!